File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Hardware.java
package org.firstinspires.ftc.teamcode;

import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
import com.qualcomm.robotcore.eventloop.opmode.OpMode;
import com.qualcomm.robotcore.hardware.CRServo;
import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.hardware.DcMotorSimple;
import com.qualcomm.robotcore.hardware.HardwareMap;
import com.qualcomm.robotcore.hardware.Servo;
import com.qualcomm.robotcore.util.ElapsedTime;

import org.firstinspires.ftc.robotcore.external.Telemetry;
import org.openftc.easyopencv.OpenCvCamera;
import org.openftc.easyopencv.OpenCvCameraRotation;
import org.openftc.easyopencv.OpenCvInternalCamera;

public class Hardware {
	private DcMotor frontLeft, frontRight,
			backLeft, backRight;
	private DcMotor[] wheels;
	private String[] wheelLabels;

	private DcMotor leftSlide, rightSlide;

	private CRServo leftSuck, rightSuck;
	private Servo leftPuller, rightPuller;

	private LinearOpMode opMode;
	private Telemetry telemetry;

	private SkystoneDetector skystoneDetector;
	private ElapsedTime timer;

	Hardware(OpMode opMode) {
		this.telemetry = opMode.telemetry;

		skystoneDetector = new SkystoneDetector(telemetry);
		timer            = new ElapsedTime();

		HardwareMap hardwareMap = opMode.hardwareMap;
		frontLeft   = hardwareMap.dcMotor.get("fl");
		frontRight  = hardwareMap.dcMotor.get("fr");
		backLeft    = hardwareMap.dcMotor.get("bl");
		backRight   = hardwareMap.dcMotor.get("br");
		leftSlide   = hardwareMap.dcMotor.get("l slide");
		rightSlide  = hardwareMap.dcMotor.get("r slide");
		leftSuck    = hardwareMap.crservo.get("ls");
		rightSuck   = hardwareMap.crservo.get("rs");
		leftPuller  = hardwareMap.servo.get("l pull");
		rightPuller = hardwareMap.servo.get("r pull");

		wheels      = new DcMotor[]{
				frontLeft, frontRight,
				backLeft, backRight
		};
		wheelLabels = new String[]{
				"FL", "FR", "BL", "BR"
		};

		frontLeft.setDirection(DcMotorSimple.Direction.REVERSE);
		backLeft.setDirection(DcMotorSimple.Direction.REVERSE);
		rightSuck.setDirection(DcMotorSimple.Direction.REVERSE);
		leftSlide.setDirection(DcMotorSimple.Direction.REVERSE);

		setWheelZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE);
		leftSlide.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE);
		rightSlide.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE);

		setWheelMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
		setWheelMode(DcMotor.RunMode.RUN_USING_ENCODER);

		telemetry.addLine("Hardware Initialized");
		telemetry.update();
	}

	void slideTele() {
		telemetry.addData("Left", "Pos: %d, Tar: %d",
				leftSlide.getCurrentPosition(), leftSlide.getTargetPosition());
		telemetry.addData("Right", "Pos: %d, Tar: %d",
				rightSlide.getCurrentPosition(), rightSlide.getTargetPosition());
		telemetry.update();
	}

	void wheelTele() {
		telemetry.addData("FL", frontLeft.getCurrentPosition());
		telemetry.addData("FR", frontRight.getCurrentPosition());
		telemetry.addData("BL", backLeft.getCurrentPosition());
		telemetry.addData("BR", backRight.getCurrentPosition());
		telemetry.update();
	}

	Hardware(LinearOpMode linearOpMode) {
		this((OpMode) linearOpMode);
		opMode = linearOpMode;
	}

	void setMecanumPower(double forwards, double strafe, double turn, double speed) {
		frontLeft.setPower((forwards - strafe + turn) * speed);
		frontRight.setPower((forwards + strafe - turn) * speed);
		backLeft.setPower((forwards + strafe + turn) * speed);
		backRight.setPower((forwards - strafe - turn) * speed);
	}

	void setLinearSlidePower(double power) {
		leftSlide.setPower(power);
		rightSlide.setPower(power);
	}

	void setSuckPower(double left, double right) {
		leftSuck.setPower(left);
		rightSuck.setPower(right);
	}

	private void setWheelMode(DcMotor.RunMode runMode) {
		for (DcMotor wheel : wheels) {
			wheel.setMode(runMode);
		}
	}

	private void setWheelZeroPowerBehavior(DcMotor.ZeroPowerBehavior behavior) {
		for (DcMotor wheel : wheels) {
			wheel.setZeroPowerBehavior(behavior);
		}
	}

	// Autonomous

	void initSkystoneDetector() {
		int cameraViewId = opMode.hardwareMap.appContext.getResources().getIdentifier(
				"cameraMonitorViewId", "id",
				opMode.hardwareMap.appContext.getPackageName());
		OpenCvCamera phoneCam =
				new OpenCvInternalCamera(OpenCvInternalCamera.CameraDirection.BACK, cameraViewId);
		phoneCam.openCameraDevice();
		phoneCam.setPipeline(skystoneDetector);
		phoneCam.startStreaming(960, 720, OpenCvCameraRotation.UPRIGHT);
	}

	int getSkystonePosition() {
		return skystoneDetector.skystonePosition();
	}

	void setPullerPositions(double left, double right) {
		leftPuller.setPosition(left);
		rightPuller.setPosition(right);
	}

	void forward(int ticks, double speed, double timeoutS) {
		move(ticks, ticks, ticks, ticks, speed, timeoutS, "Forward");
	}

	void backward(int ticks, double speed, double timeoutS) {
		move(-ticks, -ticks, -ticks, -ticks, speed, timeoutS, "Backward");
	}

	void right(int ticks, double speed, double timeoutS) {
		move(-ticks, ticks, ticks, -ticks, speed, timeoutS, "Right");
	}

	void left(int ticks, double speed, double timeoutS) {
		move(ticks, -ticks, -ticks, ticks, speed, timeoutS, "Left");
	}

	void turnRight(int ticks, double speed, double timeoutS) {
		move(ticks, -ticks, ticks, -ticks, speed, timeoutS, "Right Turn");
	}

	void turnLeft(int ticks, double speed, double timeoutS) {
		move(-ticks, ticks, -ticks, ticks, speed, timeoutS, "Left Turn");
	}

	void wait(double seconds) {
		timer.reset();
		while (timer.seconds() < seconds && opMode.opModeIsActive()) {
			telemetry.addData("Waiting", "%4.2ds");
			telemetry.update();
		}
	}

	private void move(int flTicks, int frTicks, int blTicks, int brTicks,
					  double speed, double timeoutS, String action) {
		move(flTicks, frTicks, blTicks, brTicks, speed, speed, speed, speed, timeoutS, action);
	}

	void move(int flTicks, int frTicks, int blTicks, int brTicks,
					  double flSpeed, double frSpeed, double blSpeed, double brSpeed,
					  double timeoutS, String action) {
		setWheelMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);

		frontLeft.setTargetPosition(flTicks);
		frontRight.setTargetPosition(frTicks);
		backLeft.setTargetPosition(blTicks);
		backRight.setTargetPosition(brTicks);

		setWheelMode(DcMotor.RunMode.RUN_TO_POSITION);

		frontLeft.setPower(flSpeed);
		frontRight.setPower(frSpeed);
		backLeft.setPower(blSpeed);
		backRight.setPower(brSpeed);

		timer.reset();
		while (wheelsBusy() && timer.seconds() < timeoutS && opMode.opModeIsActive()) {
			telemetry.addLine(action + "\n");
			telemetry.addData("Motor", "Position |  Target  | Distance");
			for (int i = 0; i < wheels.length; i++) {
				telemetry.addData(
						wheelLabels[i],
						"%8d | %8d | %8d",
						wheels[i].getCurrentPosition(),
						wheels[i].getTargetPosition(),
						wheels[i].getTargetPosition() - wheels[i].getCurrentPosition()
				);
			}
			telemetry.update();
		}
	}

	private boolean wheelsBusy() {
		for (DcMotor wheel : wheels) {
			if (Math.abs(wheel.getTargetPosition() - wheel.getCurrentPosition()) > 50) {
				return true;
			}
		}
		return false;
	}
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Hardware.kt
package org.firstinspires.ftc.teamcode

import com.qualcomm.hardware.bosch.BNO055IMU
import com.qualcomm.hardware.bosch.BNO055IMUImpl
import com.qualcomm.hardware.bosch.JustLoggingAccelerationIntegrator
import com.qualcomm.robotcore.hardware.*
import com.qualcomm.robotcore.hardware.DcMotor.RunMode
import com.qualcomm.robotcore.hardware.DcMotor.ZeroPowerBehavior
import com.qualcomm.robotcore.hardware.DcMotorSimple.Direction
import edu.spa.ftclib.internal.controller.ErrorTimeThresholdFinishingAlgorithm
import edu.spa.ftclib.internal.controller.FinishableIntegratedController
import edu.spa.ftclib.internal.controller.PIDController
import edu.spa.ftclib.internal.drivetrain.HeadingableMecanumDrivetrain
import edu.spa.ftclib.internal.sensor.IntegratingGyroscopeSensor
import org.firstinspires.ftc.robotcore.external.Telemetry
import org.firstinspires.ftc.robotcore.external.navigation.*
import kotlin.math.PI
import kotlin.math.atan2
import kotlin.math.hypot

open class Hardware(hardwareMap: HardwareMap, protected val telemetry: Telemetry) {
    val controller: FinishableIntegratedController
    val drivetrain: HeadingableMecanumDrivetrain

    val leftIMU: BNO055IMUImpl
    val rightIMU: BNO055IMUImpl

    val frontLeft: DcMotor
    val frontRight: DcMotor
    val backLeft: DcMotor
    val backRight: DcMotor
    val wheels: Array<DcMotor>
    val wheelLabels: Array<String>

    private val slide: DcMotor

    private val leftSuck: CRServo
    private val rightSuck: CRServo
    private val clawSlide: CRServo
    private val leftPuller: Servo
    private val rightPuller: Servo
    private val claw: Servo
    private val flicker: Servo


    init {
        with(hardwareMap) {
            frontLeft = dcMotor["fl"]
            frontRight = dcMotor["fr"]
            backLeft = dcMotor["bl"]
            backRight = dcMotor["br"]
            slide = dcMotor["l slide"]
            leftSuck = crservo["l suck"]
            rightSuck = crservo["r suck"]
            clawSlide = crservo["c slide"]
            leftPuller = servo["l pull"]
            rightPuller = servo["r pull"]
            claw = servo["claw"]
            flicker = servo["flick"]
            leftIMU = get(BNO055IMUImpl::class.java, "l imu")
            rightIMU = get(BNO055IMUImpl::class.java, "r imu")
        }

        leftIMU.initialize(BNO055IMU.Parameters().apply {
            angleUnit = BNO055IMU.AngleUnit.RADIANS
            accelUnit = BNO055IMU.AccelUnit.METERS_PERSEC_PERSEC
            loggingEnabled = true
            loggingTag = "Left IMU"
            accelerationIntegrationAlgorithm = JustLoggingAccelerationIntegrator()
        })

        rightIMU.initialize(BNO055IMU.Parameters().apply {
            angleUnit = BNO055IMU.AngleUnit.RADIANS
            accelUnit = BNO055IMU.AccelUnit.METERS_PERSEC_PERSEC
            loggingEnabled = true
            loggingTag = "Right IMU"
            accelerationIntegrationAlgorithm = JustLoggingAccelerationIntegrator()
        })

        while (!leftIMU.isGyroCalibrated || !rightIMU.isGyroCalibrated) {
            telemetry.addLine("Calibrating Gyros")
            telemetry.update()
        }

        val pid = PIDController(0.7, 0.0, 0.0)
            .apply { maxErrorForIntegral = 0.002 }

        val averageGyro = object : IntegratingGyroscope {
            override fun getAngularVelocityAxes(): MutableSet<Axis> =
                leftIMU.angularVelocityAxes

            override fun getAngularVelocity(unit: AngleUnit?): AngularVelocity {
                val left = leftIMU.getAngularVelocity(AngleUnit.RADIANS)
                val right = rightIMU.getAngularVelocity(AngleUnit.RADIANS)

                return AngularVelocity(
                    AngleUnit.RADIANS,
                    (left.xRotationRate + right.xRotationRate) / 2,
                    (left.yRotationRate + right.yRotationRate) / 2,
                    (left.zRotationRate + right.zRotationRate) / 2,
                    (left.acquisitionTime + right.acquisitionTime) / 2
                )
            }

            override fun getAngularOrientation(
                reference: AxesReference?,
                order: AxesOrder?,
                angleUnit: AngleUnit?
            ): Orientation {
                val left = leftIMU.getAngularOrientation(
                    AxesReference.EXTRINSIC,
                    AxesOrder.ZYX,
                    AngleUnit.RADIANS
                )
                val right = rightIMU.getAngularOrientation(
                    AxesReference.EXTRINSIC,
                    AxesOrder.ZYX,
                    AngleUnit.RADIANS
                )

                return Orientation(
                    AxesReference.EXTRINSIC,
                    AxesOrder.ZYX,
                    AngleUnit.RADIANS,
                    (left.firstAngle + right.firstAngle) / 2,
                    (left.secondAngle + right.secondAngle) / 2,
                    (left.thirdAngle + right.thirdAngle) / 2,
                    (left.acquisitionTime + right.acquisitionTime) / 2
                )
            }

            override fun getAngularOrientationAxes(): MutableSet<Axis> =
                leftIMU.angularOrientationAxes
        }

        controller = FinishableIntegratedController(
            IntegratingGyroscopeSensor(averageGyro),
            pid,
            ErrorTimeThresholdFinishingAlgorithm(PI / 50, 1.0)
        )

        wheels = arrayOf(frontLeft, frontRight, backLeft, backRight)
        wheelLabels = arrayOf("FL", "FR", "BL", "BR")

        drivetrain = HeadingableMecanumDrivetrain(wheels, controller)

        slide.direction = Direction.REVERSE
        rightSuck.direction = Direction.REVERSE

        wheels.forEach {
            it.zeroPowerBehavior = ZeroPowerBehavior.BRAKE

            it.mode = RunMode.STOP_AND_RESET_ENCODER
            it.mode = RunMode.RUN_USING_ENCODER
//            it.mode = RunMode.RUN_WITHOUT_ENCODER
        }

        slide.zeroPowerBehavior = ZeroPowerBehavior.BRAKE

        telemetry.addLine("Hardware Initialized")
        telemetry.update()
    }

    internal fun setLeftPullerPosition(position: PullerPosition) {
        setLeftPullerPosition(
            when (position) {
                PullerPosition.UP -> 1.0
                PullerPosition.DOWN -> 0.42
            }
        )
    }

    internal fun setLeftPullerPosition(position: Double) {
        leftPuller.position = position
    }

    internal fun setRightPullerPosition(position: PullerPosition) {
        setRightPullerPosition(
            when (position) {
                PullerPosition.UP -> 0.0
                PullerPosition.DOWN -> 0.55
            }
        )
    }

    internal fun setRightPullerPosition(position: Double) {
        rightPuller.position = position
    }

    internal fun setFlickerPosition(position: Double) {
        flicker.position = position
    }

    internal fun setClawPosition(position: Double) {
        claw.position = position
    }

    internal fun setClawSlidePower(power: Double) {
        clawSlide.power = power
    }

    internal fun setMecanumPower(
        forwards: Double,
        strafe: Double,
        turn: Double,
        speed: Double,
        reverse: Boolean
    ) {
        val forwards0 = if (reverse) -forwards else forwards
        val strafe0 = if (reverse) -strafe else strafe

        frontLeft.power = -(forwards0 - strafe0 + turn) * speed
        frontRight.power = (forwards0 + strafe0 - turn) * speed
        backLeft.power = -(forwards0 + strafe0 + turn) * speed
        backRight.power = (forwards0 - strafe0 - turn) * speed
    }

    fun setMecanumPowerGyro(
        forwards: Double,
        strafe: Double,
        turn: Double,
        speed: Double,
        reverse: Boolean
    ) {
        drivetrain.course = atan2(-forwards, strafe) + (PI / 2)
        if (reverse) {
            drivetrain.course += PI
            drivetrain.course %= 2 * PI
        }

        if (turn != 0.0) {
            drivetrain.rotation = -turn
            drivetrain.targetHeading = drivetrain.currentHeading % (2 * PI)
        } else {
            drivetrain.updateHeading()
        }

        drivetrain.velocity = hypot(strafe, forwards) * speed
    }

    internal fun setLinearSlidePower(power: Double) {
        slide.power = power
    }

    internal fun setSuckPower(left: Double, right: Double) {
        leftSuck.power = left
        rightSuck.power = right
    }
}

enum class PullerPosition {
    UP, DOWN
}
